conformance/runner.js

```diff
@@ -0,0 +1,321 @@
+
+/** Loads the Protobuf definition and the binary test data. */
+function loadTestCases() {
+  const protobufRoot = new protobufjs.Root();
+
+  protobufRoot.resolvePath = function(origin, target) {
+    if (/^google\/.*/.test(target)) {
+      target = path.join(googleProtoFiles(), target.substr('google/'.length));
+    }
+    return target;
+  };
+
+  const protoDefinition = protobufRoot.loadSync(
+    path.join(__dirname, 'test_definition.proto')
+  );
+
+  const binaryProtoData = require('fs').readFileSync(
+    path.join(__dirname, 'test_data.binprotos')
+  );
+
+  const testType = protoDefinition.lookupType('tests.Test');
+
+  const testCases = [];
+
+  for (let offset = 0; offset < binaryProtoData.length; ) {
+    const messageLength = varint.decode(binaryProtoData, offset);
+    offset += varint.encodingLength(messageLength);
+
+    let testCase = testType.decode(
+      binaryProtoData.slice(offset, offset + messageLength)
+    );
+    testCases.push(testCase);
+
+    offset = offset + messageLength;
+  }
+
+  return testCases;
+}
+
+/** Converst a JSON object into a JS Object suitable for the Node API. */
+function convertInput(json) {
+  const obj = JSON.parse(json);
+
+  function convertValue(value) {
+    if (is.object(value)) {
+      return convertObject(value);
+    } else if (is.array(value)) {
+      return convertArray(value);
+    } else if (value === 'Delete') {
+      return Firestore.FieldValue.delete();
+    } else if (value === 'ServerTimestamp') {
+      return Firestore.FieldValue.serverTimestamp();
+    }
+
+    return value;
+  }
+
+  function convertArray(arr) {
+    for (let i = 0; i < arr.length; ++i) {
+      arr[i] = convertValue(arr[i]);
+    }
+    return arr;
+  }
+
+  function convertObject(obj) {
+    for (const key in obj) {
+      if (obj.hasOwnProperty(key)) {
+        obj[key] = convertValue(obj[key]);
+      }
+    }
+    return obj;
+  }
+
+  return convertValue(obj);
+}
+
+/** Converts a CommitRequest in Proto3 JSON to Protobuf JS format. */
+function convertCommit(commitRequest) {
+  const deepCopy = JSON.parse(JSON.stringify(commitRequest));
+  for (let write of deepCopy.writes) {
+    if (write.update) {
+      write.update.fields = convert.documentFromJson(write.update.fields);
+    }
+  }
+
+  return deepCopy;
+}
+
+/** Converts a Protobuf Precondition to the expected user input. */
+function convertPrecondition(precondition) {
+  const result = JSON.parse(JSON.stringify(precondition));
+  if (result.updateTime) {
+    result.lastUpdateTime = DocumentSnapshot.toISOTime(result.updateTime);
+    delete result.updateTime;
+  }
+  return result;
+}
+
+/** List of test cases that are ignored. */
+const ignored = [
+  // Firestore Node does not omit empty writes
+  'ServerTimestamp alone',
+  'nested ServerTimestamp field',
+  'multiple ServerTimestamp fields',
+  'ServerTimestamp with dotted field',
+
+  // Test data doesn't allow update with {a:delete}
```

Comment:
The actual data in the test is `{a: {b: delete}}`. I thought the way update worked is that the top-level keys of the map or object turned into the update mask, and the values of the map/object were actual firestore values (possibly with serverTimestamp). If you wanted to delete `a.b`, you would write `{a.b: delete}`. But in `{a: {b: delete}}`, the value of `a` is supposed to be an ordinary firestore value, so it can't contain the delete sentinel.

Reply:
There are two tests that share this name (which is another reason I am a big fan of adding the method name). I believe that the request in `update-paths-11.textproto` should succeed:

```
description: "Delete cannot be nested"
update_paths: <
  doc_ref_path: "projects/projectID/databases/(default)/documents/C/d"
  field_paths: <
    field: "a"
  >
  json_values: "{\"b\": \"Delete\"}"
  is_error: true
>
```

Or does this error in Go because the field path is not included in the data?

Reply:
To me, 
```
update("a", {b: delete})
```
is the same thing as
```
update({a: {b: delete}})
```
The varargs and map/object forms should behave the same.

Reply:
Sorry, I've been misreading this test case all along. I somehow thought it was applying {b:delete()} to the document and not to the field path `a`. You are correct that this should be rejected.